float32 diff_cx
float32 diff_cy
